vignettes/benchmarks.Rmd.orig

```diff
@@ -0,0 +1,914 @@
+extract_results <- function(x, variable) {
+  stopifnot(
+    "variable must be one of c(\"timing\", \"R\", \"infections\", \"reports\")" =
+      variable %in% c("timing", "R", "infections", "reports")
+  )
+  # Return NA if there's an error
+  if (!is.null(x$error)) {
+    return(NA)
+  }
+
+  if (variable == "timing") {
+    return(round(as.duration(x$result$timing), 1))
+  } else {
+    obj <- x$result$estimates$fit
+  }
+
+  # Extracting "Rt", "infections", and "reports" is different based on the object's class and
+  # other settings
+  if (inherits(obj, "stanfit")) {
+    # Depending on rt_opts(use_rt = TRUE/FALSE), R shows up as R or gen_R
+    if (variable == "R") {
+      # The non-mechanistic model returns "gen_R" where as the others sample "R".
+      if ("R[1]" %in% names(obj)) {
+        return(extract(obj, "R")$R)
+      } else {
+        return(extract(obj, "gen_R")$gen_R)
+      }
+    } else {
+      return(extract(obj, variable)[[variable]])
+    }
+  } else {
+    obj_mat <- as_draws_matrix(obj)
+    # Extracting R depends on the value of rt_opts(use_rt = )
+    if (variable == "R") {
+      if ("R[1]" %in% variables(obj_mat)) {
+          return(subset_draws(obj_mat, "R"))
+      } else {
+        return(subset_draws(obj_mat, "gen_R"))
+      }
+    } else {
+        return(subset_draws(obj_mat, variable))
+      }
+    }
+}
+
+# Wrapper for extracting the results and making them into a data.table
+get_model_results <- function(results_by_snapshot, variable) {
+  # Get model results list
+  purrr::map_depth(results_by_snapshot, 2, extract_results, variable)
+}
+# Function to categorise fitting methods into mcmc and approximate
+add_fit_type <- function(data){
+  data[, fit_type := ifelse(fitting == "mcmc", "mcmc", "approximate")]
+  data[]
+}
+
+# Function to convert all columns to factor except the specified cols in `except` (which should be DateTime)
+make_cols_factors <- function(data, except){
+  # Make all columns except timing a factor
+  data[
+    ,
+    (setdiff(names(data), except)) :=
+      lapply(.SD, as.factor),
+    .SDcols = setdiff(names(data), except)
+  ]
+  data[]
+}
+
+# Add factor levels
+add_epidemic_phase_levels <- function(data){
+  # Add epidemic phase as a factor
+  data[, epidemic_phase := factor(epidemic_phase, levels = c("growth", "peak", "decline"))]
+  data[]
+}
+
+# Function to calculate the CRPS of log-transformed values
+calc_crps <- function(estimates, truth) {
+  # if the object is not a matrix, then it's an NA (failed run)
+  if (!inherits(estimates, c("matrix"))) return(rep(NA_real_, length(truth)))
+  # Assumes that the estimates object is structured with the samples as rows
+  shortest_obs_length <- min(ncol(estimates), length(truth))
+  reduced_truth <- tail(truth, shortest_obs_length)
+  estimates_transposed <- t(estimates) # transpose to have samples as columns
+  reduced_estimates <- tail(estimates_transposed, shortest_obs_length)
+  return(
+    crps_sample(
```

Comment:
if you're only using `crps_sample()` from `scoringutils` would it make sense to use `scoringRules` instead? As the `scoringutils` version is just a thin wrapper.

Reply:
dude

Reply:
(how are we going to goose downloads with that approach ;) )

Reply:
More seriously I think this indicates we should just use more of the features of scoringutils

Reply:
Yep, ultimately related also to https://github.com/epiforecasts/EpiNow2/issues/618